libs/Microsoft.MixedReality.WebRTC.Native/test/video_transceiver_tests.cpp

```diff
@@ -0,0 +1,634 @@
+    mrsRemoteVideoTrackHandle track_handle_remote{};
+    ASSERT_EQ(Result::kSuccess, mrsTransceiverGetRemoteVideoTrack(
+                                    transceiver_handle1, &track_handle_remote));
+    ASSERT_EQ(nullptr, track_handle_remote);
+  }
+
+  // Connect #1 and #2
+  pair.ConnectAndWait();
+
+  // The transceiver is created in its desired state, and peer #1 creates the
+  // offer, so there is no event for updating the state due to a local
+  // description.
+  ASSERT_FALSE(state_updated1_ev_local.IsSignaled());
+
+  // Wait for transceiver to be updated; this happens *after* connect,
+  // during SetRemoteDescription().
+  ASSERT_TRUE(state_updated1_ev_remote.WaitFor(10s));
+  state_updated1_ev_remote.Reset();
+
+  // Check video transceiver #1 consistency
+  {
+    // Desired state is Send+Receive, negotiated is Send only because the remote
+    // peer refused to send (no track added for that).
+    ASSERT_EQ(mrsTransceiverOptDirection::kSendOnly, dir_negotiated1);
+    ASSERT_EQ(mrsTransceiverDirection::kSendRecv, dir_desired1);
+  }
+
+  // Set transceiver #1 direction to Receive
+  ASSERT_EQ(Result::kSuccess,
+            mrsTransceiverSetDirection(transceiver_handle1,
+                                       mrsTransceiverDirection::kRecvOnly));
+  ASSERT_TRUE(state_updated1_ev_setdir.IsSignaled());
+  state_updated1_ev_setdir.Reset();
+
+  // Check video transceiver #1 consistency
+  {
+    // Desired state is Receive, negotiated is still Send only
+    ASSERT_EQ(mrsTransceiverOptDirection::kSendOnly,
+              dir_negotiated1);  // no change
+    ASSERT_EQ(mrsTransceiverDirection::kRecvOnly, dir_desired1);
+  }
+
+  // Renegotiate once the previous exchange is done
+  ASSERT_TRUE(pair.WaitExchangeCompletedFor(5s));
+  pair.ConnectAndWait();
+
+  // Wait for transceiver to be updated; this happens *after* connect, during
+  // SetRemoteDescription()
+  // Note: here the local description doesn't generate a state updated event
+  // because the local state was set with SetDirection() so is already correct.
+  // When the peer is creating the offer (#1), the desired direction is exactly
+  // the one advertized in the local description.
+  ASSERT_FALSE(state_updated1_ev_local.IsSignaled());
+  ASSERT_TRUE(state_updated1_ev_remote.WaitFor(10s));
+  state_updated1_ev_remote.Reset();
+
+  // Check video transceiver #1 consistency
+  {
+    // Desired state is Receive, negotiated is Inactive because remote peer
+    // refused to send (no track added for that).
+    ASSERT_EQ(mrsTransceiverOptDirection::kInactive, dir_negotiated1);
+    ASSERT_EQ(mrsTransceiverDirection::kRecvOnly, dir_desired1);
+  }
+
+  // Clean-up
+  mrsTransceiverRemoveRef(transceiver_handle1);
+}
+
+TEST_F(VideoTransceiverTests, SetDirection_InvalidHandle) {
+  ASSERT_EQ(
+      Result::kInvalidNativeHandle,
+      mrsTransceiverSetDirection(nullptr, mrsTransceiverDirection::kRecvOnly));
+}
+
+TEST_P(VideoTransceiverTests, SetLocalTrackSendRecv) {
```

Comment:
This and the following test are almost exactly the same, it would make sense to make the direction a parameter, or at least extract the common code into utility functions.

Reply:
Fixed.